File: src/differ.py
import fitz
import difflib

# RESET = '\033[39;49m'
# RED = '\x1b[38;5;1m'
# GREEN = '\x1b[38;5;2m'
# BLUE = '\x1b[38;5;4m'
RESET = ''
RED = ''
GREEN = ''
BLUE = ''


def open_pdf(pdf_path=None, pdf_stream=None):
    if pdf_path:
        return fitz.open(pdf_path)
    elif pdf_stream:
        return fitz.open(stream=pdf_stream)


def extract_text_from_pdf(document):
    text = ""
    for page_num in range(document.page_count):
        text += f"Page number: {page_num}\n"
        page = document.load_page(page_num)
        text += page.get_text("text")
    return text


def compare_texts(text1, text2):
    diff = difflib.ndiff(
        text1.splitlines(),
        text2.splitlines(),
    )
    return list(diff)


def leftpad(line_num):
    spaces = 4
    line_num = str(line_num)
    spaces -= len(line_num)
    return " " * spaces + line_num


def summarize_changes(diff):
    summary = []
    count = 0
    for line in diff:
        count += 1
        if line[:1] in "-":
            summary.append(
                f"{RED}{line[:1]}{leftpad(count)}:{line[1:]}{RESET}"
            )
        elif line[:1] == "+":
            summary.append(
                f"{GREEN}{line[:1]}{leftpad(count)}:{line[1:]}{RESET}"
            )
        else:
            summary.append(
                f"{BLUE} {leftpad(count)}:{line[1:]}{RESET}"
            )
    return summary


def compare_pdfs(pdf1_input, pdf2_input):
    if isinstance(pdf1_input, str):
        document1 = open_pdf(pdf1_input)
    else:
        document1 = open_pdf(pdf_stream=pdf1_input)
    if isinstance(pdf2_input, str):
        document2 = open_pdf(pdf2_input)
    else:
        document2 = open_pdf(pdf_stream=pdf2_input)

    pdf1_text = extract_text_from_pdf(document1)
    pdf2_text = extract_text_from_pdf(document2)

    diff = compare_texts(pdf1_text, pdf2_text)
    full_summary = summarize_changes(diff)
    return full_summary


if __name__ == "__main__":
    pdf1_path = 'doc1.pdf'
    pdf2_path = 'doc2.pdf'
    summary = compare_pdfs(pdf1_path, pdf2_path)
    for line in summary:
        print(line)

File: ghi/src/differ/differ.js
import { PDFDocument } from 'pdf-lib';
import { readFileSync } from 'fs';
import { diffLines } from 'diff';

async function extractTextFromPdf(pdfPath) {
    const existingPdfBytes = readFileSync(pdfPath);
    const pdfDoc = await PDFDocument.load(existingPdfBytes);
    const pages = pdfDoc.getPages();
    let text = "";

    for (let i = 0; i < pages.length; i++) {
        text += `Page number: ${i}\n`;
        const page = pages[i];
        const pageText = await page.getTextContent();
        const pageTextItems = pageText.items.map(item => item.str).join('\n');
        text += pageTextItems;
    }

    return text;
}

function compareTexts(text1, text2) {
    const diff = diffLines(text1, text2, {newlineIsToken:true});
    return diff;
}

function leftpad(line) {
    const spaces = 4;
    let num = line;
    let result = num.toString();
    while (result.length < spaces) {
        result = " " + result;
    }
    return result;
}

function summarizeChanges(diff) {
    const summary = [];
    let count = -2;
    diff.forEach((part, index) => {
        if (part.added || part.removed) {
            const lines = part.value.split('\n');
            lines.forEach(line => {
                count += 1;
                if (line) {
                    summary.push((part.added ? '+' : '-') + leftpad(count) + `:${line}`);
                }
            });
        } else {
            const lines = part.value.split('\n');
            lines.forEach(line => {
                count += 1;
                if (line) {
                    summary.push(" " + leftpad(count) + `:${line}`);
                }
            });
        }
    });
    return summary;
}

async function comparePdfs(pdf1Path, pdf2Path) {
    const pdf1Text = await extractTextFromPdf(pdf1Path);
    const pdf2Text = await extractTextFromPdf(pdf2Path);

    const diff = compareTexts(pdf1Text, pdf2Text);
    const fullSummary = summarizeChanges(diff);
    return fullSummary;
}

// Example usage
(async () => {
    const pdf1Path = 'test 4350.pdf';
    const pdf2Path = 'test 5350.pdf';
    const summary = await comparePdfs(pdf1Path, pdf2Path);
    summary.forEach(line => {
        console.log(line);
    });
})();
